### DISCUSSION TITLE
Teminology

### DISCUSSION THREAD

**Discussion by Jazcash** [beyond-all-reason/tachyon#9] [General]
Going forward I think it'd be a good idea to standardise the terminology used in the protocol. Words like `client`, `lobby`, `host` or even `game` can be confusing depending on the context, so I'd like to clear up that ambiguity in regards to this protocol at least.

My proposals:

| Term| Meaning |
|--------|--------|
| Lobby | The client application. e.g. BAR Lobby, SpringLobby, Chobby |
| Server | The provider of the protocol and what clients connect to. i.e. the master server |
| UserClient | A registered user. Can be offline or online |
| Bot | An automated UserClient, marked as a bot |
| Battle | Either custom or matchmaking. Ingame status is irrelevant from a protocol perspective |
| Founder | The user hosting a battle. Typically these users are bots |
| Host | The machine which the founder is using to serve the battle |
| Game | A version of the core game files, e.g. `Beyond All Reason test-22425-e6c0e37` |
| Engine | A version of the engine (Recoil), e.g. `105.1.1-1544-g058c8ea BAR105` |
| AI | User-owned game AI instances, e.g. ChickensAI. Can only exist within battles |

Just to be clear, I don't think these terms necessarily need to be strictly adhered to in either the server or the lobby codebases. For example, in BAR Lobby the term `bot` actually means an instance of an AI, and I believe currently in the server battles are known as `lobbies`. I think that's fine. The terminology in this protocol should be distinct from that and shouldn't be concerned with terms that are irrelevant to it.

**Comment by Teifion** [comment]
There are some issues here because I while it doesn't have to be the same on the server, it can't conflict.

Lobby refers to the period prior to a game starting and is separate from a game (battle). I've been calling Chobby a client application or lobby client.

Client is the connected representation of a user, you can't have a client without user data (but you can have a user without client data if they're not connected).

Bot vs AI is a good distinction to make, I agree with it completely.

For Host I'd rather swap to founder as host can refer to the machine/instance hosting the game itself and that might be an important distinction to make at a later stage.

**Reply by Jazcash** [reply]
Yeah you're right about users/clients. I'll change those definitions.

Think I agree about host too, could be confusing otherwise.

My one contention is on the lobby/client/lobby client/battle/game stuffs, these terminiologies as they're currently used can be quite confusing, which is why I like the word "battle" as it's less ambiguous as to what it refers to. Having a separate word for a battle that hasn't started and a battle that has started confuses things further imo, because they're essentially the same entity as far as lobby is concerned.

I think this is a similar issue to how "users" and "clients" are separated in the current tachyon implementation. As far as the lobby is concerned, these are exactly the same entity and their relationship is 1:1, so having separate commands for each makes things very messy on my end.

**Reply by Teifion** [reply]
> Lobby/Client  

If you can agree on Application for the thing the user runs that might make it easier to discuss the lobby/battle terms as it removes one of the terms we're discussing.

> Users / Clients  

Right now yes, but once we start to have more offline functions (e.g. profiles) then I think it's an important distinction to have (though I suspect you'll just say user has an online/offline flag).

**Reply by Jazcash** [reply]
> If you can agree on Application for the thing the user runs that might make it easier to discuss the lobby/battle terms as it removes one of the terms we're discussing.

How about menu? Think I'd be fine with "Lobby Application" or "Client Application" also but people will definitely shorten those to lobby and client - guess it's not super important

> Right now yes, but once we start to have more offline functions (e.g. profiles) then I think it's an important distinction to have (though I suspect you'll just say user has an online/offline flag).

That's exactly what I'll say 😛

**Reply by Teifion** [reply]
Menu describes a visual component to me, not an app. Given people will often refer to an application by the name of the app itself I still think it's an okay compromise (e.g. they'd say BAR-Lobby or Jazclient).

I'm that good at predicting. Unfortunately they have to be separate on the server so while we can combine them to UserClient I'm really not thrilled about having differing definitions (as neither will you!).

**Reply by Teifion** [reply]
@Jazcash so it pings you

**Reply by Jazcash** [reply]
Menu is just what I think people will naturally end up calling it eventually. In DOTA2 or CSGO I call it the menu also. But otherwise I think I'd be ok with calling the `Client Application` just the `Client` or the `Lobby`, if those terms don't show up anywhere else in the protocol in reference to other things.

This might be arrogant/greedy, but I'd argue the protocol schema we define here should prioritise simplicity and semanticity for clients, i.e. consumers of the protocol, rather than the server. Tbh I think I'd just rather remove the word `Client` from this protocol's dictionary completely because I don't think it's necessary in any requests/responses anyway, everything should be in reference to `Users`.

**Reply by Teifion** [reply]
> Menu

I'm surprised at that, I'd have been convinced the menu is something you show. I don't think we need to reference any menus in the protocol (it's an interface choice). I'd rather not use menu but I'm happy to concede on it if you feel strongly.

> Lobby

You wait around in a lobby and it may or may not result in a game. Given the data structure is fundamentally different I'd like to keep a distinction between games/matches and lobbies.

> User vs Client

Can we use User for an identifier of an account and ClientUser to point to a user who may or may not be logged in? Right now we'd be using the latter but I'm confident in the future we will want the former.

**Reply by Teifion** [reply]
@Jazcash

**Reply by Jazcash** [reply]
Regarding Menu/Lobby/Client/Application or whatever, I agree it's probably not important - I don't think it's gonna come up much apart from maybe the initial login command where it currently sends things like `lobby_name`, `lobby_version` and `lobby_hash`.

The `lobby` vs `game` vs `battle` thing and_the `user` vs `client` thing is something I feel quite strongly about, because I don't think the structure in this protocol redesign _should_ be fundamentally different from the client's perspective.

A battle is the same entity as an ingame battle, as a user is the same entity as an offline user - it's just they have a slightly different state, which I think would be better represented as sub-data, e.g.

```json5
{
    userId: 10,
    username: "Bob",
    rank: 12,
    status: "offline",
    battleStatus: null
}

{
    userId: 10,
    username: "Bob",
    rank: 12,
    status: "singleplayer",
    battleStatus: null
}

{
    userId: 10,
    username: "Bob",
    rank: 12,
    status: "battle",
    battleStatus: {
        battleId: 123,
        ready: false,
        teamId: 1
    }
}
```

> Given the data structure is fundamentally different
I don't quite get this - currently ingame battles have exactly the same data as pregame battles as far as protocol clients are concerned no?

My concern with separating 1:1 entities in the protocol responses is that protocol-clients are very often (if not always) going to want to keep these entities tightly coupled, and having to request each entity separately and handle the responses separately makes things a lot messier and more prone to these related entity datas from being desynced.

That being said, I understand the technical reasons for wanting to keep these entities separate from a server perspective, and if it's awkward for the server to split entities in requests from the clients or merge entities in responses to the clients then I'd be willing to compromise in the sake of progress.

So in terms of that compromise, and to bring it back to terminology, I'd be happy enough with `UserClient` being the entity that represents the combination of both user and client data, e.g.

request:
```json
{
    "cmd": "account.list_userclients",
    "data": {
        "ids": [12, 34, 56]
    }
}
```

response:
```json
{
    "cmd": "account.list_userclients",
    "data": {
        "userclients": [
            {
                "user": {
                    "user_id": 12,
                    "username": "Bob",
                    "rank": 30
                },
                "client": null
            },
            {
                "user": {
                    "user_id": 34,
                    "username": "Tim",
                    "rank": 30
                },
                "client": {
                    "battle_id": 101,
                    "ready": false,
                    "team_id": 1
                }
            }
        ]
    }
}
```

So long as all requests and responses are only for UserClients then it shouldn't be too difficult for client applications to keep these synced.

Might be good to get some second opinions in here too btw.

**Reply by Teifion** [reply]
> Regarding Menu/Lobby/Client/Application or whatever, I agree it's probably not important - I don't think it's gonna come up much apart from maybe the initial login command where it currently sends things like lobby_name, lobby_version and lobby_hash.

Given it is an application running on the user's computer are you thus happy with application? It's implied you are but I want to ensure we're on the same page. I will proceed with discussions assuming you are but it's not in the protocol until you say so.

> A battle is the same entity as an ingame battle

Not quite. There are 3 stages to a game: pre-game (lobby) in-game (moving units around) and post-game (stats, rating changes etc). There are a number of differences in both how the server and how applications would interact with each of these 3 states. I'm happy to have an overall term to cover all 3 but they're still at least 2 different types of data in 3 states so I see it as important to differentiate them.

> Client vs User

As discussed in the discord we can combine them, the majority of the server-related concerns were over having to query data in a way that made no sense. For pure nomeclature reasons and to help prevent confusion when I translate from server-internals to protocol I'd like to use the name UserClient if possible which I think you're okay with based on the above.

**Reply by Jazcash** [reply]
> Given it is an application running on the user's computer are you thus happy with application? It's implied you are but I want to ensure we're on the same page. I will proceed with discussions assuming you are but it's not in the protocol until you say so.

yep ok. so these properties would change to `app_name`, `app_version`, `app_hash`? seems fine to me

> Not quite. There are 3 stages to a game: pre-game (lobby) in-game (moving units around) and post-game (stats, rating changes etc). There are a number of differences in both how the server and how applications would interact with each of these 3 states. I'm happy to have an overall term to cover all 3 but they're still at least 2 different types of data in 3 states so I see it as important to differentiate them.

Right, but in terms of the data what are we thinking? 'Cause right now the data is the same for all 3 states:

```typescript
export type BattleOptions = {
    title: string;
    engineVersion: string;
    gameVersion: string;
    map: string;
    startPosType: StartPosType;
    startBoxes: Record<number, StartBox>;
    startTime: Date | null;
    gameOptions: Record<string, string | number | boolean>;
    mapOptions: Record<string, string | number | boolean>;
    restrictions: Restriction[];
};

export type OfflineBattleOptions = BattleOptions;

export type SpadsBattleOptions = BattleOptions & {
    id: number;
    isHost: boolean;
    passworded: boolean;
    password: string | null;
    scriptPassword: string | null;
    founderId: number;
    ip: string | null;
    port: number | null;
    locked: boolean;
    maxPlayers: number;
    preset: string;
    joinQueueUserIds: number[];
    autoBalance: string;
    teamSize: number;
    nbTeams: number;
    balanceMode: string;
};
```

A battle in the "pregame" state has all these properties and no more. A battle in the "ingame" state has all these properties and potential additional properties we may wish to add (none right now), and a battle in the "postgame" state still has all these properties and again might have some additional properties we wish to add.

Semantically, these are all the same entity. If you're saying the server has some extra info about ingame battles, then I would suggest the same thing as I did for Users and Clients, and for it to be part of the battle response if a battle is ingame/postgame:

```json5
{
    cmd: "battleUpdated",
    data: {
        battleId: 123,
        ingameStatus: {
            usersPlaying: [12, 34, 56],
            usersResigned: [78],
            gameFrame: 123456
        }
    }
}
```
```json5
{
    cmd: "battleUpdated",
    data: {
        battleId: 123,
        postgameStatus: {
            winningTeam: 1,
            statistics: [
                {
                    userId: 12,
                    damageDealt: 9999
                }
            ]
        }
    }
}
```

> As discussed in the discord we can combine them, the majority of the server-related concerns were over having to query data in a way that made no sense. For pure nomeclature reasons and to help prevent confusion when I translate from server-internals to protocol I'd like to use the name UserClient if possible which I think you're okay with based on the above.

I'd much prefer just "User", because the concept of "Clients" to represent online user data is only relevant on the server.

**Reply by Teifion** [reply]
> yep ok. so these properties would change to app_name, app_version, app_hash? seems fine to me

Yes. I'll add it to the PR :)

> Right, but in terms of the data what are we thinking? 'Cause right now the data is the same for all 3 states:

I'd be happy to have lobbydata, ingamedata and postgamedata be split into 3 chunks if it helps. But if you query a given game at each of the 3 stages you will have quite different properties. E.g. the password is not kept post-game and statistics can only exist post game.

**Comment by Jazcash** [comment]
> But if you query a given game at each of the 3 stages you will have quite different properties.

All the properties that I posted in the first snippet would be relevant in all states though no? A battle in the postgame state would still have the same password? My intention with the sub-objects (like `pregameStatus`, `ingameStats`, `postgameStatus` was that these would be the changing parts. Alternatively it might be even cleaner to just intersect the additional state-specific data based on the `state`, e.g.

```json5
{
    cmd: "battleUpdated",
    data: {
        battleId: 123,
        state: "pregame"
    }
}
```
```json5
{
    cmd: "battleUpdated",
    data: {
        battleId: 123,
        state: "ingame",
        start_time: 1682898825
    }
}
```
```json5
{
    cmd: "battleUpdated",
    data: {
        battleId: 123,
        state: "postgame",
        winning_team_id: 1
    }
}
```
If we wanted to do this, it would be easy to define these additional properties in the schema based on the literal `state`.

**Reply by Teifion** [reply]
I've thought about it a bit more. I'd like to split `lobby` from `game` or `match` completely as they are different things and both will exist at the same time. A lobby persists between games but games always need to live in a lobby and will not always exist.

Currently the server has very little info about ongoing games so this should be pretty easy to handle, the _vast_ majority of what we'd want to handle will just sit in lobbies; even saying it's in-game/post-game would be an attribute of a lobby.

**Reply by Jazcash** [reply]
Sounds sensible, I think. 1 lobby can have many matches (prefer `match` over `game` here), but it's worth noting that in the context of current state, 1 lobby can only have 1 active match, so I think it's fine to nest the current match's properties within the lobby object.

I can see we're torn on `lobby` vs `battle` terminology. With the context of lobbies being separate to matches, I think I'm more open to coming around to the `lobby` side. But with matchmaking, "lobby" doesn't really make sense.

I think in the current landscape, these terms and separations make sense, but I'm not sure if this is desirable or not. Because lobby data doesn't necessarily reflect a match's data, a match entity would have a lot of the same properties of lobby, but would reflect the actual ingame data rather than the lobby's. For example, if a lobby has map A, the match starts on map A, but then the host/boss/democracy changes the map to B _while the match is still in progress_, the lobby map won't reflect the match's map. This would be true of almost every lobby property - once a game is launched, the match's actual properties become a snapshot of what the lobby's state was at the time of launch. This is only really applicable to custom games, wouldn't apply to matchmaking.

**Reply by Teifion** [reply]
1 `lobby` can have: 0 or 1 ongoing `matches`, 0 or more past `matches`. The first definition for lobby I find is `A hall, foyer, or waiting room at or near the entrance to a building, such as a hotel or theater.` which I think works well for this purpose. Matchmaking should mostly skip the lobby experience anyway so that won't be a big deal here.

I track lobbies and matches separately so keeping them separate won't be an issue from a query perspective. It's up to you if you want to nest the current match in the lobby data or have it as a separate thing.

**Comment by Teifion** [comment]
@Jazcash am I right in thinking these are the terms and meanings we agreed upon?

| Term |	Meaning |
| ----- |	----- |
| Client application |   The application run on a computer and used to connect to the server, may also be used to launch the executable used to play the game |
| Server |	              The provider of the protocol and what clients connect to. i.e. the master or middleware server |
| UserClient |	          A registered user. Can be offline or online |
| Bot |                An automated UserClient, marked as a bot |
| Lobby |                A room or waiting area from which a Battle can be launched, one Lobby can have multiple Battles over the course of its lifetime |
| Battle |	              Either custom or matchmaking. Ingame status is irrelevant from a protocol perspective |
| Founder |	            The user hosting a battle. Typically these users are bots |
| Host |	                The machine which the founder is using to serve the battle |
| Game |                A version of the core game files, e.g. `Beyond All Reason test-22425-e6c0e37` |
| Engine |	              A version of the engine (Recoil), e.g. `105.1.1-1544-g058c8ea BAR105` |
| AI |	                  User-owned game AI instances, e.g. ChickensAI. Can only exist within battles |

**Reply by Jazcash** [reply]
Yep I think I'm fine with those

**Reply by Teifion** [reply]
I'll PR it into a doc soon
